app/code/Magento/CatalogImportExport/Model/Import/Product/LinkProcessor.php

```diff
@@ -0,0 +1,295 @@
+        '_upsell_' => \Magento\Catalog\Model\Product\Link::LINK_TYPE_UPSELL,
+    ];
+    /**
+     * @var array
+     */
+    private $linkNameToId;
+
+    /** @var LinkFactory */
+    private $linkFactory;
+
+    /** @var Helper */
+    private $resourceHelper;
+
+    /** @var SkuProcessor */
+    private $skuProcessor;
+
+    /** @var LoggerInterface */
+    private $logger;
+
+    public function __construct(
+        LinkFactory $linkFactory,
+        Helper $resourceHelper,
+        SkuProcessor $skuProcessor,
+        LoggerInterface $logger,
+        array $linkNameToId
+    ) {
+        $this->linkFactory = $linkFactory;
+        $this->resourceHelper = $resourceHelper;
+        $this->skuProcessor = $skuProcessor;
+        $this->logger = $logger;
+
+        $this->linkNameToId = $linkNameToId;
+    }
+
+    /**
+     * Gather and save information about product links.
+     *
+     * Must be called after ALL products saving done.
+     *
+     * @return $this
+     * @throws LocalizedException
+     */
+    public function saveLinks(
+        Product $importEntity,
+        Data $dataSourceModel,
+        string $linkField
+    ) {
+        /** @var Link $resource */
+        $resource = $this->linkFactory->create();
+        $mainTable = $resource->getMainTable();
+        $positionAttrId = [];
+        $nextLinkId = $this->resourceHelper->getNextAutoincrement($mainTable);
+
+        // pre-load 'position' attributes ID for each link type once
+        foreach ($this->_linkNameToId as $linkId) {
+            $select = $importEntity->getConnection()->select()->from(
+                $resource->getTable('catalog_product_link_attribute'),
+                ['id' => 'product_link_attribute_id']
+            )->where(
+                'link_type_id = :link_id AND product_link_attribute_code = :position'
+            );
+            $bind = [':link_id' => $linkId, ':position' => 'position'];
+            $positionAttrId[$linkId] = $importEntity->getConnection()->fetchOne($select, $bind);
+        }
+        while ($bunch = $dataSourceModel->getNextBunch()) {
+            $this->processLinkBunches($importEntity, $dataSourceModel, $linkField, $bunch, $resource, $nextLinkId, $positionAttrId);
+        }
+
+        return $this;
+    }
+
+    public function addNameToIds($nameToIds)
+    {
+        $this->linkNameToId = array_merge($nameToIds, $this->linkNameToId);
+    }
+
+    /**
+     * Processes link bunches
+     *
+     * @param array $bunch
+     * @param Link $resource
+     * @param int $nextLinkId
+     * @param array $positionAttrId
+     *
+     * @return void
+     * @throws LocalizedException
+     */
+    private function processLinkBunches(
+        Product $importEntity,
+        Data $dataSourceModel,
+        string $linkField,
+        array $bunch,
+        Link $resource,
+        int $nextLinkId,
+        array $positionAttrId
+    ): void {
```

Comment:
This method should be refactored and simplified to reduce cyclomatic complexity, otherwise, it might not pass the static tests

Reply:
It is moved from the parent class; there I think it passed the static test I suppose. I'd rather not touch it this time --- can still be refactored later.